api/api-cache.php

```
@@ -0,0 +1,177 @@
+     * @param int|null $httpcode (optional)
+     * @param int|null $curl_errno (optional)
+     */
+    public function __construct(string $content, bool $curl_executed = false, $httpcode = null, $curl_errno = null)
+    {
+        $this->content = $content;
+        $this->curl_executed = $curl_executed;
+        $this->httpcode = $httpcode;
+        $this->curl_errno = $curl_errno;
+    }
+}
+
+/**
+ * Gets the first entry from the API cache that has a matching endpoint. Will not return if the entry is expired.
+ * @param string $endpoint The endpoint for the API request
+ * @param resource|CurlHandle $ch The Curl Handle to send the request
+ * @param int|null $configId (optional) The config ID for the cached response. Optional if result won't be cached
+ * @return CacheResult An object containing the response content, and some cURL info if curl_exec was called
+ */
+function get_valid_cache_entry(string $endpoint, $ch, $configId = null)
+{
+    global $db;
+    if (preg_match(IGNORED_ENDPOINTS, $endpoint)) {
+        return new CacheResult(curl_exec($ch),true,curl_getinfo($ch,CURLINFO_HTTP_CODE),curl_errno($ch));
+    }
+    $q = $db->query("SELECT
+                                * 
+                            FROM 
+                                api_cache
+                            JOIN 
+                                outgoing_request_cache_config 
+                            ON 
+                                api_cache.config_id = outgoing_request_cache_config.id 
+                            WHERE 
+                                endpoint = '{$db->real_escape_string($endpoint)}'");
+    if ($q) {
+        $r = $q->fetch_array(MYSQLI_ASSOC);
+        $ttl = $r['ttl'];
+        $now = time();
+        $retrievalTime = $r['last_successfully_retrieved'];
+        if ($retrievalTime !== null && $now - strtotime($retrievalTime . 'UTC') < $ttl) {
+            return new CacheResult($r['content']);
+        }
+    }
```

I think if `$q` is false we should probably throw an error

---

Changed in c289fad [here](https://github.com/untrobotics/website/pull/158/commits/c289fad3a479b93eff024e7c5094ab293f1e3a7b#diff-970b238464763356bddea5f27fef2cba1f6162b327686a4fe165921e8bb448e9R69-R71). Would it be better to return null or treat it as though the endpoint wasn't cached in the first place (i.e., send the cURL request and return that result)?